core/src/main/java/org/jahia/modules/modulemanager/flow/ModuleManagementFlowHandler.java

```diff
@@ -177,15 +179,31 @@ public boolean uploadModule(MultipartFile moduleFile, MessageContext context, bo
             return false;
         }
         String originalFilename = moduleFile.getOriginalFilename();
-        if (!FilenameUtils.isExtension(StringUtils.lowerCase(originalFilename), "jar")) {
-            context.addMessage(new MessageBuilder().error().source("moduleFile")
-                    .code("serverSettings.manageModules.install.wrongFormat").build());
-            return false;
-        }
         File file = null;
         try {
             file = File.createTempFile("module-", "." + StringUtils.substringAfterLast(originalFilename, "."));
             moduleFile.transferTo(file);
+
+            //Handling bundles
+            boolean canHandleFile = FilenameUtils.isExtension(StringUtils.lowerCase(originalFilename), "jar");
+            BundleContext bundleContext = FrameworkService.getBundleContext();
+            Collection<ServiceReference<ArtifactUrlTransformer>> serviceReferences = bundleContext.getServiceReferences(ArtifactUrlTransformer.class, null);
+            for (ServiceReference<ArtifactUrlTransformer> serviceReference : serviceReferences) {
+                ArtifactUrlTransformer transformer = bundleContext.getService(serviceReference);
+                if (transformer.canHandle(file)) {
+                    URL transformedURL = transformer.transform(new URL("file:" + file.getPath()));
+                    FileUtils.copyInputStreamToFile(transformedURL.openConnection().getInputStream(), file);
```

Comment:
I don't know if it's ok to use the same file - Transformer will read and write from/to the same file through multiple streams. In case of npm handler it's fine as it internally create a temporary file, not sure it's the case for any handler

Reply:
Make sense, what about if I change it to only moduleFile.transferTo(file) if there's no handle to deal with the file, in this fashion I would have a clean empty file when transform/handle it.

Something like this 

`            boolean canHandleFile = FilenameUtils.isExtension(StringUtils.lowerCase(originalFilename), "jar");
            if(canHandleFile)
                moduleFile.transferTo(file);
            
            BundleContext bundleContext = FrameworkService.getBundleContext();
            Collection<ServiceReference<ArtifactUrlTransformer>> serviceReferences = bundleContext.getServiceReferences(ArtifactUrlTransformer.class, null);
            for (ServiceReference<ArtifactUrlTransformer> serviceReference : serviceReferences) {
                ArtifactUrlTransformer transformer = bundleContext.getService(serviceReference);
                if (transformer.canHandle(file)) {
                    URL transformedURL = transformer.transform(new URL("file:" + file.getPath()));
                    FileUtils.copyInputStreamToFile(transformedURL.openConnection().getInputStream(), file);
                    canHandleFile = true;
                }
            }`

Reply:
I can create a diff file to copy the result of "transformedURL.openConnection().getInputStream()" to, so I can use this new file to install the bundle.
Im not sure if that is what you mean

Reply:
I did some refactoring that maybe is solving this issue.